pymongo/message.py

```diff
@@ -878,6 +923,305 @@ def _do_batched_op_msg(
+        return request_id, msg, to_send_ops, to_send_ns
+
+    def _start(
+        self,
+        cmd: MutableMapping[str, Any],
+        request_id: int,
+        op_docs: list[Mapping[str, Any]],
+        ns_docs: list[Mapping[str, Any]],
+    ) -> MutableMapping[str, Any]:
+        """Publish a CommandStartedEvent."""
+        cmd["ops"] = op_docs
+        cmd["nsInfo"] = ns_docs
+        self.listeners.publish_command_start(
+            cmd,
+            self.db_name,
+            request_id,
+            self.conn.address,
+            self.conn.server_connection_id,
+            self.op_id,
+            self.conn.service_id,
+        )
+        return cmd
+
+
+_OP_MSG_OVERHEAD = 1000
+
+
+def _client_construct_op_msg(
+    command_doc: bytes,
+    to_send_ops: list[Mapping[str, Any]],
+    to_send_ns: list[Mapping[str, Any]],
+    ack: bool,
+    opts: CodecOptions,
+    buf: _BytesIO,
+) -> int:
+    # Write flags
+    flags = b"\x00\x00\x00\x00" if ack else b"\x02\x00\x00\x00"
+    buf.write(flags)
+
+    # Type 0 Section
+    buf.write(b"\x00")
+    buf.write(command_doc)
+
+    # Type 1 Section for ops
+    buf.write(b"\x01")
+    size_location = buf.tell()
+    # Save space for size
+    buf.write(b"\x00\x00\x00\x00")
+    buf.write(b"ops\x00")
+    # Write all the ops documents
+    for op in to_send_ops:
+        buf.write(_dict_to_bson(op, False, opts))
+    resume_location = buf.tell()
+    # Write type 1 section size
+    length = buf.tell()
+    buf.seek(size_location)
+    buf.write(_pack_int(length - size_location))
+    buf.seek(resume_location)
+
+    # Type 1 Section for nsInfo
+    buf.write(b"\x01")
+    size_location = buf.tell()
+    # Save space for size
+    buf.write(b"\x00\x00\x00\x00")
+    buf.write(b"nsInfo\x00")
+    # Write all the nsInfo documents
+    for ns in to_send_ns:
+        buf.write(_dict_to_bson(ns, False, opts))
+    # Write type 1 section size
+    length = buf.tell()
+    buf.seek(size_location)
+    buf.write(_pack_int(length - size_location))
+
+    return length
+
+
+def _client_batched_op_msg_impl(
+    command: Mapping[str, Any],
+    operations: list[tuple[str, Mapping[str, Any]]],
+    ack: bool,
+    opts: CodecOptions,
+    ctx: _ClientBulkWriteContext,
+    buf: _BytesIO,
+) -> tuple[list[Mapping[str, Any]], list[Mapping[str, Any]], int]:
+    """Create a batched OP_MSG write for client-level bulk write."""
+
+    def _check_doc_size_limits(
+        op_type: str,
+        document: Mapping[str, Any],
+        limit: int,
+    ) -> int:
+        doc_size = len(_dict_to_bson(document, False, opts))
+        if doc_size > limit:
+            _raise_document_too_large(op_type, doc_size, limit)
+        return doc_size
+
+    max_bson_size = ctx.max_bson_size
+    max_write_batch_size = ctx.max_write_batch_size
+    max_message_size = ctx.max_message_size
+
+    command_doc = _dict_to_bson(command, False, opts)
+    # Don't include command-agnostic fields in document size calculations.
```

Comment:
Is this behavior described in the spec? What is a command-agnostic field?

Reply:
The spec says to consider the size of the `bulkWrite` command document, excluding command-agnostic / operation-agnostic fields (those that aren't specific to the `bulkWrite` command) like `lsid` and `txnNumber`. The behavior is outlined [here](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#total-message-size)

Reply:
Since we have the entire command document I don't think we should do this. Instead we should check the size of `command_doc`.

Reply:
Are there benefits to checking the size of the entire `command_doc` vs. checking the abridged size and accounting for the excluded fields in the 1000 byte overhead allowance? I was wondering because the spec mentions [here](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#how-was-the-op_msg-overhead-allowance-determined) that even if it's possible to determine the exact size of the OP_MSG, we should follow the given overhead calculations in order to standardize things across drivers

Reply:
We should still add the 1000 byte overhead, I'm only suggesting we remove the "abridged" concept since it's extra unneeded complexity.

Reply:
Ahh got it -- I think the only place the abridged concept is necessary is for a few of the prose tests about batch-splitting. For example, [this test](https://github.com/mongodb/specifications/tree/master/source/crud/tests#case-2-batch-splitting-required) relies on the command document's size being calculated based only on the bulkWrite-specific keys.

Reply:
Does that test fail if we change:
```python
    # Don't include bulkWrite-command-agnostic fields in document size calculations.
    abridged_keys = ["bulkWrite", "errorsOnly", "ordered"]
    if command.get("bypassDocumentValidation"):
        abridged_keys.append("bypassDocumentValidation")
    if command.get("comment"):
        abridged_keys.append("comment")
    if command.get("let"):
        abridged_keys.append("let")
    command_abridged = {key: command[key] for key in abridged_keys}
    command_len_abridged = len(_dict_to_bson(command_abridged, False, opts))
```
to:
```python
    command_abridged = command
    command_len_abridged = len(_dict_to_bson(command, False, opts))
```

Reply:
Yep that causes the test to fail, since the command doc size is now larger than what's expected [here](https://github.com/mongodb/specifications/tree/master/source/crud/tests#details-on-size-calculations)